File: content/LLD/Reference-Code/19. Composite Pattern/good_example.py
from typing import List
from abc import ABC, abstractmethod


class FileSystemComponent(ABC):
    @abstractmethod
    def show_details(self):
        pass


class File(FileSystemComponent):
    def __init__(self, name: str):
        self.__name = name

    def show_details(self):
        return f"File : {self.__name}"


class Folder(FileSystemComponent):
    def __init__(self, name: str):
        self.__name = name
        self.__components: List[FileSystemComponent] = []

    def add_component(self, component: FileSystemComponent):
        self.__components.append(component)

    def show_details(self):
        print(f"Folder name: {self.__name}")
        for component in self.__components:
            print(component.show_details())


file1 = File("image.png")
file2 = File("ppt")
file3 = File("word.exe")

sub_folder = Folder("sub folder")
sub_folder.add_component(file1)
sub_folder.add_component(file2)
sub_folder.add_component(file3)

image_file = File("imaaageeee")
movie = File("krissh")
main_folder = Folder("main folder")
main_folder.add_component(image_file)
main_folder.add_component(movie)
main_folder.add_component(sub_folder)

main_folder.show_details()

File: content/LLD/Reference-Code/19. Composite Pattern/GoodExample.java
import java.util.ArrayList;
import java.util.List;

public class GoodExample {
    interface FileSystemComponent {
        void showDetails();
    }

    static class File implements FileSystemComponent {
        private final String name;

        public File(String name) {
            this.name = name;
        }

        @Override
        public void showDetails() {
            System.out.println("File : " + this.name);
        }
    }

    static class Folder implements FileSystemComponent {
        private final String name;
        private final List<FileSystemComponent> components = new ArrayList<>();

        public Folder(String name) {
            this.name = name;
        }

        public void addComponent(FileSystemComponent component) {
            this.components.add(component);
        }

        @Override
        public void showDetails() {
            System.out.println("Folder name: " + this.name);
            for (FileSystemComponent component : this.components) {
                component.showDetails();
            }
        }
    }

    public static void main(String[] args) {
        File file1 = new File("image.png");
        File file2 = new File("ppt");
        File file3 = new File("word.exe");

        Folder subFolder = new Folder("sub folder");
        subFolder.addComponent(file1);
        subFolder.addComponent(file2);
        subFolder.addComponent(file3);

        File imageFile = new File("imaaageeee");
        File movie = new File("krissh");
        Folder mainFolder = new Folder("main folder");
        mainFolder.addComponent(imageFile);
        mainFolder.addComponent(movie);
        mainFolder.addComponent(subFolder);

        mainFolder.showDetails();
    }
}
